subsys/bluetooth/host/hci_core.c

```diff
@@ -671,6 +671,33 @@ int bt_hci_disconnect(uint16_t handle, uint8_t reason)
 	return bt_hci_cmd_send_sync(BT_HCI_OP_DISCONNECT, buf, NULL);
 }
 
+static uint16_t disconnected_handles[CONFIG_BT_MAX_CONN];
+static void conn_handle_disconnected(uint16_t handle)
+{
+	for (int i = 0; i < ARRAY_SIZE(disconnected_handles); i++) {
+		if (!disconnected_handles[i]) {
+			disconnected_handles[i] = 0x8000 | handle;
+		}
+	}
+}
+
+static bool conn_handle_is_disconnected(uint16_t handle)
+{
+	/* Set one bit so that 0 can be used as a valid handle, only 12 bits
+	 * is used for valid handles.
+	 */
+	handle |= 0x8000;
```

Comment:
Suggest to create a macro for this specific bit (or even setting all the top 4 bits), to avoid the "magic bit" case :D

Reply:
@Thalley I've updated the description of the commit and the PR to better explain the problem and the fix.
It's not something the controller does, but a host handling of the events, that leads to the situation.

While the fix may not look particularly pretty it was the best I could come up with in the current design while keeping the same deadlock releasing behavior.

Also added a macro for valid bits as suggested.

Reply:
Thanks - The commit message is very clear now :)